src/lib.rs

```diff
@@ -30,77 +30,84 @@ use criterion::{
     Throughput,
 };
 
-#[cfg(not(any(
-    target_arch = "x86_64",
-    target_arch = "x86",
-    target_arch = "aarch64",
-    target_arch = "loongarch64"
-)))]
-compile_error!(
-    "criterion-cycles-per-byte currently works only on x86 or x86_64 or aarch64 or loongarch64."
-);
-
 /// `CyclesPerByte` measures clock cycles using the CPU read time-stamp counter instruction. `cpb` is
 /// the preferred measurement for cryptographic algorithms.
 pub struct CyclesPerByte;
 
 // WARN: does not check for the cpu feature; but we'd panic anyway so...
-fn rdtsc() -> u64 {
-    #[cfg(target_arch = "x86_64")]
-    unsafe {
-        core::arch::x86_64::_mm_mfence();
-        core::arch::x86_64::_rdtsc()
-    }
-
-    #[cfg(target_arch = "x86")]
-    unsafe {
-        core::arch::x86::_mm_mfence();
-        core::arch::x86::_rdtsc()
-    }
-
-    #[cfg(all(target_arch = "aarch64", target_os = "linux"))]
-    unsafe {
-        // If a aarch64 CPU, running GNU/Linux kernel, executes following instruction,
-        // it'll *probably* panic with message "illegal instruction executed", because userspace
-        // isn't allowed to execute that instruction without installing a Linux Kernel Module.
-        //
-        // I've tested the LKM @ https://github.com/jerinjacobk/armv8_pmu_cycle_counter_el0
-        // on a Raspberry Pi 4b ( i.e. ARM Cortex-A72, running kernel version 6.5.0-1006-raspi )
-        // and it works like charm. While extending support of this library for aarch64 targets,
-        // I found https://github.com/pornin/crrl#benchmarks pretty helpful.
-        let mut counter: u64;
-        core::arch::asm!("dsb sy", "mrs {}, pmccntr_el0", out(reg) counter);
-        counter
-    }
-
-    #[cfg(target_arch = "loongarch64")]
+#[inline(always)]
+fn cycle_counter() -> u64 {
     unsafe {
-        let counter: u64;
-        core::arch::asm!("rdtime.d {0}, $zero", out(reg) counter);
-        counter
+        cfg_if::cfg_if! {
+            if #[cfg(all(rdpru, any(target_arch = "x86_64", target_arch = "x86")))] {
+                let ret: u64;
+                core::arch::asm!(
```

Comment:
Would it be possible to move some of these instructions  out of the `asm!` and into normal rust code? It would be great to have as little inline assembly as possible, even if that means having separate cfg blocks for rdpru on x86 and rdpru on x86_64?

Reply:
Yes, we can use the `_mm_lfence` intrinsic. AFAIK Rust `std` currently does not have an intrinsic for RDTSC. The alternative code would look like [this](https://rust.godbolt.org/z/oG17hW48E). Personally, I think it's slightly less readable, but if you prefer it, I will update PR to use it.

Reply:
Oh that's exactly what I wanted. Thank you. I agree it looks a bit uglier but I think it's a lot clearer what's going on.